FIx (xml-) config classes for thread-safe issues

304NotModified opened the pull request:
Replacing foreach with for, for thread-safe usage

`For` forces uses to use lists.

(also `for` is most of the time more performant)

fixes https://github.com/NLog/NLog/issues/1153

Page-Not-Found commented:
Wouldn't it be better if you expose a lock variable and use the same during modification of the list?

304NotModified commented:
Good question.  Well locks are more expensive then creating objects. (CPU, lock is some kind of thread sleep)  And when the collection is changed while processing, it's isn't wrong to continue progressing the old collection.  

Or do you see it otherwise?

Page-Not-Found commented:
@304NotModified I don't have any other thoughts.

Page-Not-Found commented on src/NLog/Config/LoggingConfiguration.cs:
```diff
@@ -302,8 +308,10 @@ public void Uninstall(InstallationContext installationContext)
         {
             InternalLogger.Debug("Closing logging configuration...");
-            foreach (ISupportsInitialize initialize in this.configItems.OfType<ISupportsInitialize>())
+            var supportsInitializesList = this.configItems.OfType<ISupportsInitialize>().ToList();
```
Can you create method similar to `GetInstallableItems`

304NotModified replied:
Thanks. I needed a parameters for the `reverse()` in this case (performance)

Page-Not-Found commented on src/NLog/Config/LoggingConfiguration.cs:
```diff
@@ -361,10 +369,12 @@ internal void Dump()
         {
             var uniqueTargets = new List<Target>();
-            foreach (var rule in this.LoggingRules)
+            for (int i = 0; i < this.LoggingRules.Count; i++)
```
Wouldn't we need to take a copy into a variable like we did in other places?

304NotModified replied:
Because it's already a `List` and no foreach. (so safe)

Page-Not-Found replied:
I was meaning to say When another thread is updating the list and you are trying to make a `for` loop using the `this.LoggingRules.Count` wouldn't it cause the same issue which we have currently?

304NotModified replied:
You have a good point here. We don't get the "System.InvalidOperationExceptionCollection was modified; enumeration operation may not execute.", but it's still not thread-safe

Page-Not-Found replied:
Do you think this needs to be fixed? If not can you resolve the conflicts, I can merge this.

304NotModified replied:
I will fix it. (assigned myself also) 

Funny,  do your see an conflict? A here it says there are no conflicts in this PR.

Page-Not-Found replied:
Let me check later today.
On Jan 17, 2016 2:51 PM, "Julian Verdurmen" <notifications@github.com>
wrote:

> In src/NLog/Config/LoggingConfiguration.cs
> <https://github.com/NLog/NLog/pull/1157#discussion_r49951158>:
>
> > @@ -361,10 +369,12 @@ internal void Dump()
> >          internal void FlushAllTargets(AsyncContinuation asyncContinuation)
> >          {
> >              var uniqueTargets = new List<Target>();
> > -            foreach (var rule in this.LoggingRules)
> > +            for (int i = 0; i < this.LoggingRules.Count; i++)
>
> I will fix it. (assigned myself also)
>
> Funny, do your see an conflict? A here it says there are no conflicts in
> this PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NLog/NLog/pull/1157/files#r49951158>.
>

Page-Not-Found commented on src/NLog/Config/LoggingConfiguration.cs:
```diff
@@ -397,8 +411,10 @@ internal void ValidateConfig()
             InternalLogger.Info("Found {0} configuration items", this.configItems.Length);
 
-            foreach (object o in this.configItems)
+            var configItemList = this.configItems.ToList();
```
`InternalLogger.Info("Found {0} configuration items", this.configItems.Length);`
Might need to move this here. 

Why would you want to convert a array to a list? Since the for can act on a array.

304NotModified replied:
> Why would you want to convert a array to a list? Since the for can act on a array.

This is unnecessary indeed. Thanks!

304NotModified commented:
Thanks for the review @Page-Not-Found .

I applied most of your suggestions, see code comment . Please double check. If you think it's OK, just merge it :)

Page-Not-Found commented:
Sure. I just now added some more comments. Can you also check to see there
is no merge conflicts?
On Jan 15, 2016 14:01, "Julian Verdurmen" <notifications@github.com> wrote:

> Thanks for the review @Page-Not-Found <https://github.com/Page-Not-Found>
> .
>
> I applied most of your suggestions, see code comment . Please double
> check. If you think it's OK, just merge it :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NLog/NLog/pull/1157#issuecomment-172054095>.
>

Page-Not-Found commented:
I don't see the Merge pull request button. it is disabled.

304NotModified commented:
(Travis has some issues, e.g. timeouts when restoring packages, in multiple branches seen)

304NotModified commented:
@Page-Not-Found if you check this again, please do. 

I reverted (and squashed) some previous commutes. I use now every time a variable with `.ToList()`  (which makes a new list always, see reference source)

304NotModified commented:
@Page-Not-Found, you're now a gladiator? ;)

Page-Not-Found commented:
I am good with all of the code now. But I still prefer to use the `for` construct just because it does not create a enumerator in run time. 

One small suggestion, when you have just one line with in a `if` or `for` block. Don't need `{}`

bhaeussermann commented:
> One small suggestion, when you have just one line with in a if or for block. Don't need {}

That's a matter of style ;). I personally prefer omitting `{}` when not needed (to keep the code more concise), but others like to include them always. I guess it makes single-line code blocks more visually explicit and consistent with other blocks.

304NotModified commented:
```c
if ((err = SSLHashSHA1.update(&hashCtx, &serverRandom)) != 0)
    goto fail;
if ((err = SSLHashSHA1.update(&hashCtx, &signedParams)) != 0)
    goto fail;
    goto fail;  
if ((err = SSLHashSHA1.final(&hashCtx, &hashOut)) != 0)
    goto fail;

err = sslRawVerify(...);
```

The enormous SSL fail of Apple (2014)

Page-Not-Found commented:
I agree will you both. We are good to merge.

304NotModified commented:
Thanks for checking @Page-Not-Found!
